Fix gitignore and dcc search path

JohnstonJ opened the pull request:
These commits fix a few issues:
- The Delphi search path is missing the `source` directory, which breaks incremental makes in C++Builder (but `Build` still succeeds as apparently RAD Studio builds very differently between these two options).  I added this missing directory to the search path so that incremental makes work again.
- For some reason, the repository only provides a `.hgignore` file for Mercurial.  Copy this to a `.gitignore` file to ignore project outputs.
- When recompiling LockBox using a RAD Studio edition that doesn't match that of the author, `.RES` files are needlessly changed - causing a dirty repository.  I deleted these `.RES` files from source control and added them to `.gitignore`.

JohnstonJ committed:
Fix problems with project files that prevent compiling LockBox on C++Builder XE8:

* Add source directory to DCC search path; this was breaking incremental makes.
* Remove invalid directories from include and library paths.

JohnstonJ committed:
Copied .hgignore file to create .gitignore file.

JohnstonJ committed:
Remove resource files from source control and ignore them.

RAD Studio has a tendency to modify these files, even when you are merely
building the projects from a clean repository - for example, if compiling
using a different edition of RAD Studio than the one that originally created
the file.  This obnoxiously makes the repository appear dirty.  Remove them
from source control as they contain nothing valuable.

JohnstonJ committed:
Set up project dependencies in the project group.

Previously, it was allowed to build design packages without having build the runtime package.

romankassebaum commented:
1. I added '....\source' to the search path for all Delphi and C++Builder packages.
2. I added .gitignore to the project although I'm working with hg.
3. But I'm not convinced that I should remove the .res file. I fear that it could miss if someone works with runtime packages.

JohnstonJ commented:
Thanks, see comments below:

> From: Roman Kassebaum [mailto:notifications@github.com] 
> Sent: Thursday, August 20, 2015 11:25
> To: TurboPack/LockBox
> Cc: James Johnston
> Subject: Re: [LockBox] Fix gitignore and dcc search path (#2)
> 1. I added '....\source' to the search path for all Delphi and C++Builder packages.

What of the other invalid paths that I removed?  I found these would cause compiler warnings to spew to the output. 

> 1. I added .gitignore to the project although I'm working with hg.

I don't think the "syntax: glob" at the top is not recognized in a .gitignore; I've never seen it before.  You'd want to remove it if you want a proper .gitignore file.

But I found that the rest of the .gitignore functioned fine with it in the file so it's not a big deal.

> 1. But I'm not convinced that I should remove the .res file. I fear that it could miss if someone works with runtime packages.

What's the exact scenario where this would be a problem?  I found if the .res files are deleted, they are regenerated without issue.  And more to the point, if a .res file already exists and is committed to the repository, it makes the entire repository show up as dirty if/when RAD Studio changes it - and all it takes in this case is opening the project and hitting "build."  It's quite annoying and in my case I'm adding LockBox as a submodule to be built as part of a larger project, and I'd like LockBox to keep its working/source tree clean.

Finally, what about my commit that fixed build dependencies in the project group?  This lack of dependency information is causing breakage if somebody tries to build a design-time package before the runtime-package was built.

Best regards,

James Johnston

romankassebaum commented:
> What of the other invalid paths that I removed? I found these would cause compiler warnings to spew to the output.

But there is nothing else than "....\source" in the source path.

> I don't think the "syntax: glob" at the top is not recognized in a .gitignore; I've never seen it before. You'd want to remove it if you want a proper .gitignore file.

But then .hgignore and .gitignore are out of sync.

> What's the exact scenario where this would be a problem? I found if the .res files are deleted, they are regenerated without issue.

I explained it in issue 3 with the hpp files. It is always possible that someone simply downloads LockBox without compiling it.

> It's quite annoying and in my case I'm adding LockBox as a submodule to be built as part of a larger project, and I'd like LockBox to keep its working/source tree clean.

You shouldn't recompile the third party packages in an automatic build process. If you simply compile the code then the res file is not recreated.

> Finally, what about my commit that fixed build dependencies in the project group? This lack of dependency information is causing breakage if somebody tries to build a design-time package before the runtime-package was built.

I don't like the build dependencies. It always makes the IDE slower if you are developing the designtime packages.
